### ISSUE TITLE
Combination of for_each and nested modules causes performance issues

### ISSUE THREAD

**Issue by ethankent** [bridgecrewio/checkov#6068]
**Describe the issue**
Graph generation appears to bog down when checking a collection of modules (generated using `count` or `for_each`), which include a submodule. The example below takes about 2.5 minutes to run.

**Examples**

```
# modules.tf
module "parent" {
  count = 12
  source = "./parent"
  name   = count.index
}
output "result" {
  value = { for k, v in module.parent : k => v }
}
```

```
# parent/main.tf
variable "name" {
  type = string
}
module "child" {
  source = "../child"
  name   = var.name
}
output "result" {
  value = module.child.result
}
```

```
# child/main.tf
variable "name" {
  type = string
}
resource "terraform_data" "example" {
  input = var.name
}
output "result" {
  value = terraform_data.example.output
}
```

**Desktop (please complete the following information):**
 - OS: Provided by `ghcr.io/bridgecrewio/checkov:3.2.31@sha256:98202e56dc524e6a7b4d80f0a641817e7f94130f5e8bf653f210d0203693d20c`
 - Provided by `ghcr.io/bridgecrewio/checkov:3.2.31@sha256:98202e56dc524e6a7b4d80f0a641817e7f94130f5e8bf653f210d0203693d20c`

**Additional context**
#5443 noted same/similar behavior and that a [fix](https://github.com/bridgecrewio/checkov/pull/5463/files#diff-6824483924fc6d42cae9c83ff1a31076d6358c6bf480acf5e6768725c91cf0d0R153-R156) had been made. However, this code appears to have been subsequently [removed](https://github.com/bridgecrewio/checkov/commit/840e6b32c5b30ea129ca124d2f4354a8414e0c83#diff-6824483924fc6d42cae9c83ff1a31076d6358c6bf480acf5e6768725c91cf0d0R155-R159).

**Comment by bo156** [comment]
Hi @ethankent :) 

First of all thanks for pointing that out! 
The PR's you mentioned under `additonal context` actually are both related to correct rendering and not directly for performance (might have been related indircetly).

Anyway, I was able to recreate the issue and opened a [fix for it here](https://github.com/bridgecrewio/checkov/pull/6072).
The root cause was excessive calls to `os.path.realpath` under `terraform.graph_builder.local_graph._find_vertex_with_best_match`. 

The PR I opened reduced the runtime to around 1-1.5 minutes for the code you provided.
I also suspect there is an extra issue about the amount of vertices we create in the graph which also affects performance, will keep you updated.

**Comment by bo156** [comment]
One more update - I added another commit to the mentioned PR which fixes the issue completely and now the run only takes a few seconds.

The issue was that when using nested-modules together with count index > 2, we mistakenly duplicated more resources than needed. This caused an "exponential" effect so the larger the count was we duplicated more resources, which drastically affected performance as we had a lot more resource to scan. 
Thank you for opening this issue! @ethankent
